File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if(amount == 0) return 0;
        int[] dpArray = new int[amount + 1];
        for(int i = 1; i <= amount; i++) {
            dpArray[i] = Integer.MAX_VALUE;
            for(int coin: coins) {
                if(coin <= i && dpArray[i - coin] != Integer.MAX_VALUE) {
                    dpArray[i] = Math.min(dpArray[i],1 + dpArray[i-coin]);
                }
             }
        }
        return dpArray[amount] == Integer.MAX_VALUE ? -1 : dpArray[amount];
    }

    
}

File: 322-coin-change/coin-change.py
class Solution(object):
    def coinChange(self, coins, amount):
        dp = [0] * (amount + 1)
        for i in range(1,amount + 1):
            dp[i] = float("inf")
            for coin in coins:
                if coin <= i and dp[i-coin] != float("inf"):
                    dp[i] = min(dp[i],1 + dp[i-coin])
        
        return dp[amount] if dp[amount] != float("inf") else -1